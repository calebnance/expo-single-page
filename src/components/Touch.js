import React from 'react';
import { Text, TouchableOpacity } from 'react-native';
import PropTypes from 'prop-types';
import { gStyle } from '../constants';

const Touch = ({ activeO, onPress, style, text, textStyle }) => (
  <TouchableOpacity activeOpacity={activeO} onPress={onPress} style={style}>
    <Text style={textStyle}>{text}</Text>
  </TouchableOpacity>
);

// default props
Touch.defaultProps = {
  activeO: gStyle.activeO,
  style: {},
  text: '',
  textStyle: {
    textAlign: 'center'
  }
};

// type checking
Touch.propTypes = {
  // required
  onPress: PropTypes.func.isRequired,

  // optional
  activeO: PropTypes.number,
  style: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.number,
    PropTypes.object
  ]),
  text: PropTypes.string,
  textStyle: PropTypes.object
};

export default Touch;
