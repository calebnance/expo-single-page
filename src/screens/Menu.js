import React from 'react';
import { Animated, ScrollView, StyleSheet, View } from 'react-native';
import PropTypes from 'prop-types';
import { colors, device } from '../constants';

// components
import Touch from '../components/Touch';

class Menu extends React.Component {
  constructor(props) {
    super(props);

    let positionStart = -device.width;
    if (props.direction === 'right') {
      positionStart = device.width;
    }

    this.state = {
      backgroundColor: new Animated.Value(0),
      position: new Animated.Value(positionStart),
      positionStart
    };

    this.handleClose = this.handleClose.bind(this);
  }

  componentDidMount() {
    const { backgroundColor, position } = this.state;

    Animated.timing(backgroundColor, {
      delay: 100,
      duration: 400,
      toValue: 10
    }).start();

    Animated.timing(position, {
      duration: 400,
      toValue: 0
    }).start();
  }

  handleClose() {
    const { onClose } = this.props;
    const { backgroundColor, position, positionStart } = this.state;

    Animated.timing(backgroundColor, {
      duration: 100,
      toValue: 0
    }).start();

    Animated.timing(position, {
      duration: 400,
      toValue: positionStart
    }).start(() => {
      onClose();
    });
  }

  renderLayout() {
    const { direction } = this.props;
    const { backgroundColor } = this.state;

    let render = '';
    const color = backgroundColor.interpolate({
      inputRange: [0, 10],
      outputRange: [colors.white0, colors.white20]
    });

    const mockItems = [];

    for (let i = 0; i < 20; i += 1) {
      mockItems.push(<View key={i} style={styles.mockBlock} />);
    }

    const scrollViewArea = (
      <ScrollView style={styles.scrollView}>
        {mockItems}
        <View style={styles.listSpacer} />
      </ScrollView>
    );

    const closeArea = (
      <Touch onPress={() => this.handleClose()} style={styles.closeArea} />
    );

    if (direction === 'left') {
      render = (
        <Animated.View style={[styles.panel, { backgroundColor: color }]}>
          {scrollViewArea}
          {closeArea}
        </Animated.View>
      );
    } else if (direction === 'right') {
      render = (
        <Animated.View style={[styles.panel, { backgroundColor: color }]}>
          {closeArea}
          {scrollViewArea}
        </Animated.View>
      );
    }

    return render;
  }

  render() {
    const { show } = this.props;
    const { position } = this.state;

    if (show === false) {
      return null;
    }

    return (
      <Animated.View style={[styles.container, { left: position }]}>
        {this.renderLayout()}
      </Animated.View>
    );
  }
}

// default props
Menu.defaultProps = {
  direction: 'left',
  show: false
};

// type checking
Menu.propTypes = {
  // required
  onClose: PropTypes.func.isRequired,

  // optional
  direction: PropTypes.string,
  show: PropTypes.bool
};

const styles = StyleSheet.create({
  container: {
    height: '100%',
    position: 'absolute',
    width: '100%',
    zIndex: 50
  },
  panel: {
    flex: 1,
    flexDirection: 'row'
  },
  scrollView: {
    backgroundColor: colors.brown,
    paddingTop: device.iPhoneX ? 44 : 0,
    width: '80%'
  },
  closeArea: {
    height: '100%',
    width: '20%'
  },
  mockBlock: {
    backgroundColor: colors.grey,
    height: 75,
    marginBottom: 20,
    width: '100%'
  },
  listSpacer: {
    height: device.iPhoneX ? 44 : 0
  }
});

export default Menu;
