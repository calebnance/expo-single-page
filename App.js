import React from 'react';
import { StatusBar, StyleSheet, Text, View } from 'react-native';
import { colors, device } from './src/constants';

// screens
import Menu from './src/screens/Menu';

// components
import InfoBar from './src/components/InfoBar';
import Touch from './src/components/Touch';

export default class App extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      progress: 10,
      showLeftMenu: false,
      showRightMenu: false,
      statusBar: 'light-content',
      statusBarHidd: false,
      statusBarTrans: 'fade'
    };

    this.menuClose = this.menuClose.bind(this);
    this.menuOpen = this.menuOpen.bind(this);
  }

  menuClose() {
    this.setState({
      showLeftMenu: false,
      showRightMenu: false,
      statusBar: 'light-content',
      statusBarHidd: false
    });
  }

  menuOpen(showMenu) {
    this.setState(prevState => ({
      progress: prevState.progress + 10,
      showLeftMenu: showMenu === 'showLeftMenu',
      showRightMenu: showMenu === 'showRightMenu',
      statusBar: 'dark-content',
      statusBarHidd: true
    }));
  }

  renderMenus() {
    const { showLeftMenu, showRightMenu } = this.state;
    let menuScreen = null;

    if (showLeftMenu) {
      menuScreen = <Menu show={showLeftMenu} onClose={this.menuClose} />;
    } else if (showRightMenu) {
      menuScreen = (
        <Menu direction="right" show={showRightMenu} onClose={this.menuClose} />
      );
    }

    return menuScreen;
  }

  render() {
    const { progress, statusBar, statusBarHidd, statusBarTrans } = this.state;

    return (
      <View style={styles.container}>
        <StatusBar
          backgroundColor={colors.blue}
          barStyle={statusBar}
          hidden={statusBarHidd}
          showHideTransition={statusBarTrans}
        />

        <InfoBar progress={progress} text="info bar" />

        <Text style={styles.text}>welcome!</Text>
        <Text style={styles.text}>
          this is a simple one page react native app
        </Text>

        <Touch
          style={[styles.block, styles.leftBlock]}
          testId="leftMenu"
          text="left menu"
          onPress={() => this.menuOpen('showLeftMenu')}
        />
        <Touch
          style={[styles.block, styles.rightBlock]}
          testId="rightMenu"
          text="right menu"
          onPress={() => this.menuOpen('showRightMenu')}
        />

        {this.renderMenus()}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    backgroundColor: colors.brandPrimary,
    flex: 1,
    justifyContent: 'center'
  },
  text: {
    color: colors.white
  },
  block: {
    alignItems: 'center',
    borderRadius: 4,
    bottom: device.iPhoneNotch ? 32 : 16,
    height: 64,
    padding: 4,
    position: 'absolute',
    justifyContent: 'center',
    width: 64
  },
  leftBlock: {
    backgroundColor: colors.blue,
    left: 16
  },
  rightBlock: {
    backgroundColor: colors.blue,
    right: 16
  }
});
