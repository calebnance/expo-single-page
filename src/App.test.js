import React from 'react';
import renderer from 'react-test-renderer';

import App from '../App';
import Menu from './screens/Menu';
import Touch from './components/Touch';

describe('<App />', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('renders correctly', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has 2 touch components', () => {
    const testRenderer = renderer.create(<App />);
    const testInstance = testRenderer.root;
    const touches = testInstance.findAllByType(Touch);

    expect(touches.length).toBe(2);
  });

  it('shows left menu', () => {
    const testRenderer = renderer.create(<App />);
    const testInstance = testRenderer.root;
    const touchLeftMenu = testInstance.findByProps({ testId: 'leftMenu' });

    touchLeftMenu.props.onPress();
  });

  it('shows right menu', () => {
    const testRenderer = renderer.create(<App />);
    const testInstance = testRenderer.root;
    const touchRightMenu = testInstance.findByProps({ testId: 'rightMenu' });

    touchRightMenu.props.onPress();
  });

  it('closes menu', () => {
    const testRenderer = renderer.create(<App />);
    const testInstance = testRenderer.root;
    const touchLeftMenu = testInstance.findByProps({ testId: 'leftMenu' });
    touchLeftMenu.props.onPress();

    const menuInstance = testInstance.findByType(Menu);
    menuInstance.props.onClose();
  });
});
