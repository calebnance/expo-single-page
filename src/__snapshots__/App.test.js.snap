// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#1d2d3a",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#fff",
        "borderRadius": 4,
        "overflow": "hidden",
        "paddingVertical": 8,
        "position": "absolute",
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.4,
        "shadowRadius": 1,
        "top": 44,
        "width": 200,
        "zIndex": 70,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#d7263d",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      info bar
    </Text>
    <View
      style={
        Array [
          Object {
            "bottom": 0,
            "position": "absolute",
            "width": "100%",
          },
          Object {
            "height": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.2)",
            },
          ]
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "#d7263d",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": "10%",
          }
        }
      />
    </View>
  </View>
  <Text
    style={
      Object {
        "color": "#fff",
      }
    }
  >
    welcome!
  </Text>
  <Text
    style={
      Object {
        "color": "#fff",
      }
    }
  >
    this is a simple one page react native app
  </Text>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#0197f6",
        "borderRadius": 4,
        "bottom": 16,
        "height": 64,
        "justifyContent": "center",
        "left": 16,
        "opacity": 1,
        "padding": 4,
        "position": "absolute",
        "width": 64,
      }
    }
  >
    <Text
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      left menu
    </Text>
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#0197f6",
        "borderRadius": 4,
        "bottom": 16,
        "height": 64,
        "justifyContent": "center",
        "opacity": 1,
        "padding": 4,
        "position": "absolute",
        "right": 16,
        "width": 64,
      }
    }
  >
    <Text
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      right menu
    </Text>
  </View>
</View>
`;
